master/buildbot/worker/libvirt.py

```diff
@@ -281,33 +222,31 @@ def stop_instance(self, fast=False):
         If the VM was using a cloned image, I remove the clone
         When everything is tidied up, I ask that bbot looks for work to do
         """
-        @defer.inlineCallbacks
-        def _destroy_domain(res, domain):
-            log.msg('Graceful shutdown failed. Force destroying domain {}'.format(self.workername))
-            # Don't return res to stop propagating shutdown error if destroy
-            # was successful.
-            yield domain.destroy()
+        domain = self.getDomain()
 
         log.msg("Attempting to stop '{}'".format(self.workername))
-        if self.domain is None:
+        if domain.ID() == -1:
             log.msg("I don't think that domain is even running, aborting")
-            return defer.succeed(None)
-
-        domain = self.domain
-        self.domain = None
+            return None
 
         if self.graceful_shutdown and not fast:
             log.msg("Graceful shutdown chosen for {}".format(self.workername))
-            d = domain.shutdown()
-            d.addErrback(_destroy_domain, domain)
+            with self.getLock():
```

Comment:
you should look at other latent worker for idea of implementation.

Idea for blocking code, is to use thread pool, and to call threaded method.

I would associate one connection per thread so that you know that it will not be shared.

Then you just have to build a _start_instance_thd() method, which is alone in its thread and you don't have to 
manage ressource lock.
If your connection is bound to a thread it is then not shared apart from threaded calls
The threaded call is effectively a barrier/lock